pennylane/tape/functions/unified.py

```diff
@@ -0,0 +1,210 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Wrappers for common functions that manipulate or create
+NumPy, TensorFlow, and Torch data structures.
+"""
+import abc
+
+
+class UnifiedTensor(abc.ABC):
+    """Creates a unified tensor wrapper, that represents interface array/tensor datatypes
+    using a common API.
+
+    While this is an abstract base class, this class may be 'instantiated' directly;
+    by overloading ``__new__``, the tensor argument is inspected, and the correct subclass
+    is returned:
+
+    >>> x = tf.Variable([0.4, 0.1, 0.5])
+    >>> y = UnifiedTensor(x)
+    >>> type(y)
+    <class 'pennylane.tape.functions.tf.TensorFlowTensor'>
+
+    The original tensor is available via the ``data`` attribute.
+
+    In addition, this class defines various abstract methods that all subclasses
+    must define. These methods allow for common manipulations and
+    linear algebra transformations without the need for importing.
+
+    >>> y.ones_like()
+    tf.Tensor([1. 1. 1.], shape=(3,), dtype=float32)
+
+    Unless specified, the returned tensors are also unified tensors, allowing
+    for method chaining:
+
+    >>> y.ones_like().expand_dims(0)
+    tf.Tensor([[1. 1. 1.]], shape=(1, 3), dtype=float32)
+    """
+
+    def __new__(cls, tensor):
+        if isinstance(tensor, UnifiedTensor):
+            return tensor
+
+        if cls is not UnifiedTensor:
+            return super(UnifiedTensor, cls).__new__(cls)

```

Comment:
Not sure what this is for?

Reply:
Oh, this is to avoid an infinite recursion! This way, if the class is not `UnifiedTensor`, e.g., it is a subclass, then the logic below is completely skipped, and the instantiation is allowed to proceed as normal.